### PR TITLE
Update to final version of ML-KEM, and use seed as decapsulation key.

### PR DESCRIPTION
**Pull request by bwesterb**

### CODE REVIEW

File: draft-connolly-cfrg-xwing-kem.md [hunk 1]
```diff
@@ -224,14 +240,16 @@ X-Wing relies on the following primitives:
...
 
-  - `ML-KEM-768.KeyGenDerand(seed)`: Same as `ML-KEM-768.KeyGen()`,
-    but derandomized as follows.
-    `seed[0:32]` is used for `d` in line 1 of algorithm 12 from {{MLKEM}}
-    and `seed` is 64 bytes. `seed[32:64]` is used for `z` in line 1 of
-    algorithm 15.
-  - `ML-KEM-768.EncapsDerand(pk_M, seed)`: Same as `ML-KEM-768.Encaps()`
-    but derandomized as follows.
-    `seed` is 32 bytes and used for `m` in line of 1 algorithm 16.
+  - `ML-KEM-768.KeyGen_internal(d, z)`: Algorith to generate an
```
Comment (dconnolly) [inline draft-connolly-cfrg-xwing-kem.md:243]:
```suggestion
  - `ML-KEM-768.KeyGen_internal(d, z)`: Algorithm to generate an
```

**Comment by sophieschmieg** [comment]
LGTM.

We do need a separate proof for the combined KEM, using the fact that we include the X25519 public key and ephemeral public key, otherwise we do not get HON-BIND-K-PK/HON-BIND-K-CT.
We could also shorten the list of misbinding properties to only list the MAL-BIND properties, since the others are true a fortiori.

**Comment by dconnolly** [comment]
> LGTM.
> 
> We do need a separate proof for the combined KEM, using the fact that we include the X25519 public key and ephemeral public key, otherwise we do not get HON-BIND-K-PK/HON-BIND-K-CT. We could also shorten the list of misbinding properties to only list the MAL-BIND properties, since the others are true a fortiori.

@bwesterb we should update the eprint with the MAL-BIND proof(s)

File: draft-connolly-cfrg-xwing-kem.md [hunk 2]
```diff
@@ -390,21 +412,52 @@ ciphertext `ct`.
...
+For efficiency, an implementation MAY cache the result of `expandDecapsulationKey`.
+This is useful in two cases:
+
+1. If multiple ciphertexts for the same key are decapsulated.
+2. If a ciphertext is decapsulated for a key that has just been generated.
+   This happen on the client-side for TLS.
+
+A typical API pattern to achieve this optimisation is to have an
+opaque decapsulation key object that hides the cached values.
+For instance, such an API could have the following functions.
```
Comment (dconnolly) [inline draft-connolly-cfrg-xwing-kem.md:439]: Yes this is an opinion but:

```suggestion
A typical API pattern to achieve this optimization is to have an
opaque decapsulation key object that hides the cached values.
For instance, such an API could have the following functions.
```

File: draft-connolly-cfrg-xwing-kem.md [hunk 3]
```diff
@@ -390,21 +412,52 @@ ciphertext `ct`.
...
+    and returns the packed decapsulation key.
+
+The expanded decapsulation key could cache more computation,
+such as the expanded matrix A in ML-KEM.
+
+Any such expanded decapsulation key MUST NOT be transmitted between
+implementations, as this could break the security analysis of X-Wing.
+In particular, the MAL-BIND-K-PK and MAL-BIND-K-CT binding
+properties of X-Wing do not hold when transmitting the regular ML-KEM
+decapsulation key.
```
Comment (dconnolly) [inline draft-connolly-cfrg-xwing-kem.md:459]: <3

File: draft-connolly-cfrg-xwing-kem.md [hunk 4]
```diff
@@ -460,6 +513,26 @@ combiner cannot be assumed to be secure, when used with different
...
+- LEAK-BIND-K-CT
+- LEAK-BIND-K,CT-PK
+- LEAK-BIND-K,PK-CT
+- HON-BIND-K-PK
+- HON-BIND-K-CT
+- HON-BIND-K,CT-PK
+- HON-BIND-K,PK-CT
+
+In contrast, ML-KEM on its own does not achieve
+MAL-BIND-K-PK, MAL-BIND-K-CT, nor MAL-BIND-K,PK-CT. {{SCHMIEG}}
```
Comment (dconnolly) [inline draft-connolly-cfrg-xwing-kem.md:534]: Yeah it looks like they didn't change the algorithms or keygen sufficiently to affect this

**Review by dconnolly — APPROVED** [review]

**Pull request merged**